### PR TITLE
[Optimization] Update estimated_num_new_pages logic in TokenToKVPoolAllocator

### PR DESCRIPTION
**Pull request by YiXR**
## Motivation

This PR update estimated_num_new_pages logic and format code which is introduced by the previous PR: https://github.com/sgl-project/sglang/pull/8133

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/mem_cache/allocator.py [hunk 1]
```diff
@@ -79,6 +79,9 @@ def free_group_end(self):
         if self.free_group:
             self.free(torch.cat(self.free_group))
 
+    def estimated_num_new_pages(self, bs, extend_num_tokens):
+        return bs * ((extend_num_tokens + self.page_size - 1) // self.page_size)
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/allocator.py:85]: Maybe add a note to clarify the logic here.

**Review by ShangmingCai — APPROVED** [review]
LGTM.

File: python/sglang/srt/mem_cache/allocator.py (hunk 1 above)
Comment (YiXR) [reply python/sglang/srt/mem_cache/allocator.py:85]: Here I use upper bound `bs * ((extend_num_tokens + self.page_size - 1) // self.page_size)`
instead of accurate value `((seq_lens + self.page_size - 1) // self.page_size - (prefix_lens + self.page_size - 1) // self.page_size).sum().item()` 
to  reduce the computational overhead caused by each estimate. The increase in the frequency of reordering caused by using the upper bound is very small and can be ignored.

**Review by merrymercy — COMMENTED** [review]
Looks good to me.
It seems this sort is only useful for PD case. Should we add a flag so that it only turns on during PD?

**Comment by YiXR** [comment]
> Looks good to me. It seems this sort is only useful for PD case. Should we add a flag so that it only turns on during PD?

done

**Review by xiezhq-hermann — APPROVED** [review]

File: python/sglang/srt/mem_cache/allocator.py [hunk 2]
```diff
@@ -433,8 +453,9 @@ def alloc(self, need_size: int):
             ), "The allocation size should be page-aligned"
 
         num_pages = need_size // self.page_size
-        if num_pages > len(self.free_pages):
-            self.merge_and_sort_free()
+        if self.need_sort:
+            if num_pages > len(self.free_pages):
+                self.merge_and_sort_free()
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/allocator.py]: nit: should we use `and` instead of another `if`?

Comment (YiXR) [reply python/sglang/srt/mem_cache/allocator.py]: fixed
